File: backend/src/services/tenant.service.ts
import { v4 as uuidv4 } from 'uuid';
import db from '../utils/database';
import keycloakService from './keycloak.service';
import logger from '../utils/logger';
import { auditService } from './audit.service';

/**
 * Tenant data structure
 */
export interface Tenant {
  id: string;
  name: string;
  slug: string;
  keycloak_realm: string;
  schema_name: string;
  status: 'active' | 'suspended' | 'pending';
  settings: Record<string, any>;
  created_at: Date;
  updated_at: Date;
}

/**
 * Tenant creation request
 */
export interface CreateTenantRequest {
  name: string;
  slug: string;
  adminEmail: string;
  adminFirstName: string;
  adminLastName: string;
  adminPassword: string;
  settings?: Record<string, any>;
}

/**
 * Tenant Service
 *
 * Handles all tenant lifecycle operations including:
 * - Provisioning new tenants (DB schema + Keycloak realm)
 * - Tenant configuration management
 * - Tenant deletion/cleanup
 *
 * This is the core multi-tenancy orchestration layer.
 */
class TenantService {
  /**
   * Provisions a new tenant with complete isolation.
   *
   * This operation:
   * 1. Creates a tenant record in the platform schema
   * 2. Creates an isolated database schema for the tenant
   * 3. Creates a Keycloak realm with roles and client
   * 4. Creates the initial admin user
   */
  async createTenant(request: CreateTenantRequest): Promise<Tenant> {
    const tenantLogger = logger.child({ operation: 'createTenant', slug: request.slug });
    tenantLogger.info('Starting tenant provisioning');

    // Generate identifiers
    const tenantId = uuidv4();
    const schemaName = `tenant_${request.slug.toLowerCase().replace(/[^a-z0-9]/g, '_')}`;
    const realmName = request.slug.toLowerCase().replace(/[^a-z0-9]/g, '-');

    try {
      // Step 1: Create tenant record in platform schema
      tenantLogger.info('Creating tenant record');
      const result = await db.query<Tenant>(
        `INSERT INTO platform.tenants (id, name, slug, keycloak_realm, schema_name, status, settings)
         VALUES ($1, $2, $3, $4, $5, 'pending', $6)
         RETURNING *`,
        [
          tenantId,
          request.name,
          request.slug,
          realmName,
          schemaName,
          JSON.stringify(request.settings || {}),
        ]
      );

      const tenant = result.rows[0];

      // Step 2: Create database schema for tenant data isolation
      tenantLogger.info('Creating tenant database schema');
      await db.createTenantSchema(schemaName);

      // Step 3: Create Keycloak realm
      tenantLogger.info('Creating Keycloak realm');
      await keycloakService.createRealm(realmName, request.name);

      // Step 4: Create roles in the realm
      tenantLogger.info('Creating realm roles');
      await keycloakService.createRealmRoles(realmName);

      // Step 5: Create client application
      tenantLogger.info('Creating Keycloak client');
      await keycloakService.createClient(realmName);

      // Step 6: Create admin user
      tenantLogger.info('Creating admin user');
      const adminKeycloakId = await keycloakService.createUser(
        realmName,
        request.adminEmail,
        request.adminFirstName,
        request.adminLastName,
        request.adminPassword,
        ['admin']
      );

      // Step 7: Sync admin user to tenant database
      tenantLogger.info('Syncing admin user to tenant database');
      const adminUserId = uuidv4();
      await db.queryWithTenant(
        schemaName,
        `INSERT INTO users (id, keycloak_id, email, first_name, last_name, role, status)
         VALUES ($1, $2, $3, $4, $5, 'admin', 'active')`,
        [adminUserId, adminKeycloakId, request.adminEmail, request.adminFirstName, request.adminLastName]
      );

      // Step 8: Activate tenant
      tenantLogger.info('Activating tenant');
      await db.query(
        `UPDATE platform.tenants SET status = 'active', updated_at = NOW() WHERE id = $1`,
        [tenantId]
      );

      // Log successful provisioning for audit
      await auditService.log({
        tenantId,
        action: 'TENANT_CREATED',
        resourceType: 'tenant',
        resourceId: tenantId,
        details: {
          name: request.name,
          slug: request.slug,
          adminEmail: request.adminEmail,
        },
      });

      tenantLogger.info('Tenant provisioning completed successfully');

      return {
        ...tenant,
        status: 'active',
      };
    } catch (error) {
      tenantLogger.error('Tenant provisioning failed', { error });

      // Attempt cleanup on failure
      try {
        await this.cleanupFailedProvisioning(tenantId, schemaName, realmName);
      } catch (cleanupError) {
        tenantLogger.error('Cleanup after failed provisioning also failed', { cleanupError });
      }

      throw error;
    }
  }

  /**
   * Cleans up resources after a failed provisioning attempt.
   */
  private async cleanupFailedProvisioning(
    tenantId: string,
    schemaName: string,
    realmName: string
  ): Promise<void> {
    logger.warn('Cleaning up failed tenant provisioning', { tenantId, schemaName, realmName });

    // Delete tenant record
    await db.query('DELETE FROM platform.tenants WHERE id = $1', [tenantId]);

    // Drop schema if created
    try {
      await db.dropTenantSchema(schemaName);
    } catch (e) {
      // Schema might not exist
    }

    // Delete realm if created
    try {
      await keycloakService.deleteRealm(realmName);
    } catch (e) {
      // Realm might not exist
    }
  }

  /**
   * Gets a tenant by ID.
   */
  async getTenantById(id: string): Promise<Tenant | null> {
    const result = await db.query<Tenant>(
      'SELECT * FROM platform.tenants WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  /**
   * Gets a tenant by slug.
   */
  async getTenantBySlug(slug: string): Promise<Tenant | null> {
    const result = await db.query<Tenant>(
      'SELECT * FROM platform.tenants WHERE slug = $1',
      [slug]
    );
    return result.rows[0] || null;
  }

  /**
   * Gets a tenant by Keycloak realm name.
   */
  async getTenantByRealm(realm: string): Promise<Tenant | null> {
    const result = await db.query<Tenant>(
      'SELECT * FROM platform.tenants WHERE keycloak_realm = $1',
      [realm]
    );
    return result.rows[0] || null;
  }

  /**
   * Lists all tenants with optional filtering.
   */
  async listTenants(options?: {
    status?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ tenants: Tenant[]; total: number }> {
    const conditions: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (options?.status) {
      conditions.push(`status = $${paramIndex++}`);
      params.push(options.status);
    }

    const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    // Get total count
    const countResult = await db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM platform.tenants ${whereClause}`,
      params
    );

    // Get tenants with pagination
    const limit = options?.limit || 50;
    const offset = options?.offset || 0;

    const result = await db.query<Tenant>(
      `SELECT * FROM platform.tenants ${whereClause}
       ORDER BY created_at DESC
       LIMIT $${paramIndex++} OFFSET $${paramIndex}`,
      [...params, limit, offset]
    );

    return {
      tenants: result.rows,
      total: parseInt(countResult.rows[0].count, 10),
    };
  }

  /**
   * Updates tenant settings.
   */
  async updateTenantSettings(
    tenantId: string,
    settings: Record<string, any>
  ): Promise<Tenant> {
    const result = await db.query<Tenant>(
      `UPDATE platform.tenants
       SET settings = settings || $1, updated_at = NOW()
       WHERE id = $2
       RETURNING *`,
      [JSON.stringify(settings), tenantId]
    );

    if (result.rows.length === 0) {
      throw new Error('Tenant not found');
    }

    await auditService.log({
      tenantId,
      action: 'TENANT_SETTINGS_UPDATED',
      resourceType: 'tenant',
      resourceId: tenantId,
      details: { settings },
    });

    return result.rows[0];
  }

  /**
   * Suspends a tenant (disables access).
   */
  async suspendTenant(tenantId: string, reason: string): Promise<void> {
    const tenant = await this.getTenantById(tenantId);
    if (!tenant) {
      throw new Error('Tenant not found');
    }

    await db.query(
      `UPDATE platform.tenants SET status = 'suspended', updated_at = NOW() WHERE id = $1`,
      [tenantId]
    );

    await auditService.log({
      tenantId,
      action: 'TENANT_SUSPENDED',
      resourceType: 'tenant',
      resourceId: tenantId,
      details: { reason },
    });

    logger.warn('Tenant suspended', { tenantId, reason });
  }

  /**
   * Reactivates a suspended tenant.
   */
  async reactivateTenant(tenantId: string): Promise<void> {
    await db.query(
      `UPDATE platform.tenants SET status = 'active', updated_at = NOW() WHERE id = $1`,
      [tenantId]
    );

    await auditService.log({
      tenantId,
      action: 'TENANT_REACTIVATED',
      resourceType: 'tenant',
      resourceId: tenantId,
    });

    logger.info('Tenant reactivated', { tenantId });
  }

  /**
   * Deletes a tenant and all associated data.
   * WARNING: This is destructive and irreversible.
   */
  async deleteTenant(tenantId: string): Promise<void> {
    const tenant = await this.getTenantById(tenantId);
    if (!tenant) {
      throw new Error('Tenant not found');
    }

    logger.warn('Deleting tenant', { tenantId, name: tenant.name });

    // Delete Keycloak realm
    try {
      await keycloakService.deleteRealm(tenant.keycloak_realm);
    } catch (error) {
      logger.error('Failed to delete Keycloak realm', { error });
    }

    // Drop database schema
    await db.dropTenantSchema(tenant.schema_name);

    // Delete tenant record
    await db.query('DELETE FROM platform.tenants WHERE id = $1', [tenantId]);

    await auditService.log({
      action: 'TENANT_DELETED',
      resourceType: 'tenant',
      resourceId: tenantId,
      details: { name: tenant.name, slug: tenant.slug },
    });
  }
}

export const tenantService = new TenantService();
export default tenantService;

File: admin-service/app/services/tenant_service.py
from uuid import UUID, uuid4
from typing import Optional
import logging
import json
import re

from app import database as db
from app.models.tenant import CreateTenantRequest, TenantResponse
from app.services.keycloak_service import keycloak_service

logger = logging.getLogger(__name__)


class TenantService:
    """Service for tenant lifecycle management."""

    async def create_tenant(self, request: CreateTenantRequest) -> TenantResponse:
        """
        Provision a new tenant with:
        1. Database record
        2. Keycloak realm
        3. Database schema
        4. Admin user (in both Keycloak and DB)
        """
        tenant_id = uuid4()
        schema_name = f"tenant_{re.sub(r'[^a-z0-9]', '_', request.slug.lower())}"
        realm_name = request.slug  # Use slug as realm name

        logger.info(f"Creating tenant: {request.name} ({request.slug})")

        # Step 1: Insert tenant record with pending status
        await db.execute(
            """
            INSERT INTO platform.tenants (id, name, slug, schema_name, keycloak_realm, status, settings)
            VALUES ($1, $2, $3, $4, $5, 'pending', $6)
            """,
            tenant_id,
            request.name,
            request.slug,
            schema_name,
            realm_name,
            json.dumps(request.settings or {}),
        )

        try:
            # Step 2: Create Keycloak realm
            await keycloak_service.create_realm(realm_name, request.name)

            # Step 3: Create database schema
            await db.create_tenant_schema(schema_name)

            # Step 4: Create admin user in Keycloak
            keycloak_user_id = await keycloak_service.create_user(
                realm_name=realm_name,
                email=request.admin_email,
                first_name=request.admin_first_name,
                last_name=request.admin_last_name,
                password=request.admin_password,
                roles=["admin"],
            )

            # Step 5: Create admin user in tenant database
            user_id = uuid4()
            async with db.get_connection() as conn:
                await conn.execute(f'SET search_path TO "{schema_name}"')
                await conn.execute(
                    """
                    INSERT INTO users (id, email, first_name, last_name, password_hash, role, status)
                    VALUES ($1, $2, $3, $4, $5, 'admin', 'active')
                    """,
                    user_id,
                    request.admin_email,
                    request.admin_first_name,
                    request.admin_last_name,
                    keycloak_user_id,  # Store Keycloak ID instead of password hash
                )

            # Step 6: Activate tenant
            await db.execute(
                """
                UPDATE platform.tenants
                SET status = 'active', updated_at = NOW()
                WHERE id = $1
                """,
                tenant_id,
            )

            logger.info(f"Tenant created successfully: {tenant_id}")

            # Fetch and return the created tenant
            return await self.get_tenant(tenant_id)

        except Exception as e:
            logger.error(f"Failed to create tenant: {e}")
            # Cleanup on failure
            await self._cleanup_failed_tenant(tenant_id, schema_name, realm_name)
            raise

    def _hash_password(self, password: str) -> str:
        """Simple password hashing (use bcrypt in production)."""
        import hashlib
        return hashlib.sha256(password.encode()).hexdigest()

    async def _cleanup_failed_tenant(
        self, tenant_id: UUID, schema_name: str, realm_name: str
    ) -> None:
        """Clean up resources after a failed tenant creation."""
        try:
            # Delete Keycloak realm
            try:
                await keycloak_service.delete_realm(realm_name)
            except Exception as e:
                logger.error(f"Failed to delete Keycloak realm: {e}")

            # Drop schema
            try:
                await db.drop_tenant_schema(schema_name)
            except Exception as e:
                logger.error(f"Failed to drop schema: {e}")

            # Delete tenant record
            await db.execute("DELETE FROM platform.tenants WHERE id = $1", tenant_id)

        except Exception as e:
            logger.error(f"Cleanup failed: {e}")

    async def list_tenants(
        self,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> tuple[list[TenantResponse], int]:
        """List tenants with optional filtering and pagination."""
        where_clause = ""
        params = []

        if status:
            where_clause = "WHERE status = $1"
            params.append(status)

        # Get total count
        count_query = f"SELECT COUNT(*) FROM platform.tenants {where_clause}"
        total = await db.fetchval(count_query, *params)

        # Get tenants
        query = f"""
            SELECT t.*, (
                SELECT COUNT(*)
                FROM information_schema.tables
                WHERE table_schema = t.schema_name AND table_name = 'users'
            ) as has_users
            FROM platform.tenants t
            {where_clause}
            ORDER BY t.created_at DESC
            LIMIT ${len(params) + 1} OFFSET ${len(params) + 2}
        """
        params.extend([limit, offset])

        rows = await db.fetch(query, *params)

        tenants = []
        for row in rows:
            tenant = TenantResponse(
                id=row["id"],
                name=row["name"],
                slug=row["slug"],
                status=row["status"],
                schema_name=row["schema_name"],
                settings=json.loads(row["settings"]) if row["settings"] else {},
                created_at=row["created_at"],
                updated_at=row["updated_at"],
            )

            # Get user count if schema exists
            if row["has_users"]:
                try:
                    count = await db.fetchval(
                        f'SELECT COUNT(*) FROM "{row["schema_name"]}".users'
                    )
                    tenant.user_count = count
                except Exception:
                    tenant.user_count = 0

            tenants.append(tenant)

        return tenants, total

    async def get_tenant(self, tenant_id: UUID) -> Optional[TenantResponse]:
        """Get a tenant by ID."""
        row = await db.fetchrow(
            "SELECT * FROM platform.tenants WHERE id = $1", tenant_id
        )

        if not row:
            return None

        tenant = TenantResponse(
            id=row["id"],
            name=row["name"],
            slug=row["slug"],
            status=row["status"],
            schema_name=row["schema_name"],
            settings=json.loads(row["settings"]) if row["settings"] else {},
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )

        # Get user count
        try:
            count = await db.fetchval(
                f'SELECT COUNT(*) FROM "{row["schema_name"]}".users'
            )
            tenant.user_count = count
        except Exception:
            tenant.user_count = 0

        return tenant

    async def get_tenant_by_slug(self, slug: str) -> Optional[TenantResponse]:
        """Get a tenant by slug."""
        row = await db.fetchrow(
            "SELECT * FROM platform.tenants WHERE slug = $1", slug
        )

        if not row:
            return None

        return await self.get_tenant(row["id"])

    async def update_tenant(
        self, tenant_id: UUID, name: Optional[str] = None, settings: Optional[dict] = None
    ) -> Optional[TenantResponse]:
        """Update tenant name and/or settings."""
        updates = []
        params = []
        param_idx = 1

        if name is not None:
            updates.append(f"name = ${param_idx}")
            params.append(name)
            param_idx += 1

        if settings is not None:
            updates.append(f"settings = ${param_idx}")
            params.append(json.dumps(settings))
            param_idx += 1

        if not updates:
            return await self.get_tenant(tenant_id)

        updates.append("updated_at = NOW()")
        params.append(tenant_id)

        query = f"""
            UPDATE platform.tenants
            SET {', '.join(updates)}
            WHERE id = ${param_idx}
        """

        await db.execute(query, *params)
        return await self.get_tenant(tenant_id)

    async def suspend_tenant(self, tenant_id: UUID, reason: str) -> None:
        """Suspend a tenant."""
        await db.execute(
            """
            UPDATE platform.tenants
            SET status = 'suspended', updated_at = NOW()
            WHERE id = $1
            """,
            tenant_id,
        )
        logger.info(f"Suspended tenant {tenant_id}: {reason}")

    async def reactivate_tenant(self, tenant_id: UUID) -> None:
        """Reactivate a suspended tenant."""
        await db.execute(
            """
            UPDATE platform.tenants
            SET status = 'active', updated_at = NOW()
            WHERE id = $1
            """,
            tenant_id,
        )
        logger.info(f"Reactivated tenant {tenant_id}")

    async def delete_tenant(self, tenant_id: UUID) -> None:
        """Permanently delete a tenant and all its data."""
        tenant = await db.fetchrow(
            "SELECT * FROM platform.tenants WHERE id = $1", tenant_id
        )

        if not tenant:
            raise ValueError("Tenant not found")

        logger.info(f"Deleting tenant: {tenant_id}")

        # Delete Keycloak realm
        if tenant.get("keycloak_realm"):
            try:
                await keycloak_service.delete_realm(tenant["keycloak_realm"])
            except Exception as e:
                logger.error(f"Failed to delete Keycloak realm: {e}")

        # Drop database schema
        try:
            await db.drop_tenant_schema(tenant["schema_name"])
        except Exception as e:
            logger.error(f"Failed to drop schema: {e}")

        # Delete tenant record
        await db.execute("DELETE FROM platform.tenants WHERE id = $1", tenant_id)

        logger.info(f"Tenant deleted: {tenant_id}")


tenant_service = TenantService()
